### ISSUE TITLE
Allow serving of a "default asset"

### ISSUE THREAD

**Issue by jroper** [playframework/playframework#7634]
A typical SPA these days will do all its own routing, using the browser history APIs to update URLs. In this case, a router for Play app that has an SPA frontend typically needs to do the following:

* Serve any REST calls that it has implemented
* Serve the SPA assets from a particular path
* On all other routes, serve the SPAs index.html, allowing it to handle all routing, 404s etc.

This is not straight forward to implement though, the reason being is that often there's no automatic way to determine whether a URL is for an SPA asset or not from the router. A solution to this would be to make it possible to define a "default asset". When defined, the assets router would, when it can't find an asset, serve that asset instead. And for these SPAs, the default asset would be the index.html.  It might look like this:

```
GET  /*file     Assets.at(path = "/public", file, defaultAsset = "index.html")
```

**Comment by richdougherty** [comment]
We could also implement this by putting some hooks into the `Assets` class so that users can subclass it and override its behaviour when no asset is found.

**Comment by gmethvin** [comment]
I like @richdougherty's idea of having a hook to override the default behavior. We can have a protected `assetNotFound` method in `AssetsBuilder`. If I'm rendering a single-page app, I might also be interested in rendering some initial state on the server, or a template, or any number of different things as the default.

Another possibility is to add something to configuration, for example:

```
play.assets.default = "index.html"
```

where the default value would be `null`, meaning we should render the default error page. Or we could do a combination of both.

**Comment by domdorn** [comment]
which status code would you serve here? still 404? else people might get problems with google+seo+duplicate content, e.g. when some urls where generated wrong in one version of the app and now they cannot get them out of the index because the assets controller returns a 2xx code for urls actually not existing

**Comment by gmethvin** [comment]
The default should certainly still be 404, but depending on how your SPA is implemented you might not have a way to know on the server side if that path is valid, so it might be more desirable to serve a 200 and deal with the nonexistent page on the client side.

**Comment by mrubin** [comment]
From the perspective of a new user (of Play) and knowing nothing about the implementation, I absolutely vote for a "default asset" approach. I shouldn't have to subclass something when this is a 3-word one-liner in nginx. My expectation would be that I could achieve this in the routes file only.

If people have a need to subclass, let them. But make what should be a simple thing easy (and allow hard things to be possible)  :)

**Comment by mrubin** [comment]
As far as the http status, if index.html (or wherever the default asset is) gets successfully served up, then that should absolutely be a 200. 

Greg is correct that the server has no way of knowing what routes exist in the SPA. The front-end has to deal with that.

**Comment by gmethvin** [comment]
Maybe it would be best to have something explicit:

```
GET /*file  Assets.at(path = "/public", file, fallback = "index.html", fallbackStatus = OK)
```

**Comment by richdougherty** [comment]
Can we clarify what sort of "default asset" we're talking about…

Are we talking about the situation where a user requests `/foo/bar/` and we serve them `/foo/bar/index.html`? I think this is what @mrubin is talking about and I agree it's a very common case.

Or are we talking about an asset that's served whenever there's a 404, e.g. a user requests `/foo/bar/no-such-asset.html` and gets `/index.html`? This might be what @jroper's talking about.

**Comment by marcospereira** [comment]
@richdougherty if I understand it correctly, both @jroper and @mrubin are talking about the same use case: SPA that does some kind of client-side routing and then serving some default asset when the required route does not exist.

This is not about serving an `index.html` there is present inside `/foo/bar` folder.

**Comment by mrubin** [comment]
@richdougherty You're close.

The nginx logic of `try_files $uri /index.html` says: "If I can serve what the user requested, I will. Else, serve up the contents of /index.html (without rewriting the url in the browser)."

So in our case, if the routes file said:

`GET /*file  Assets.at(path = "/public", file, fallback = "index.html", fallbackStatus = OK)`

There's a bit of ambiguity as to the path to look at for `index.html`. My assumption would be that we look at `/public/index.html`. So if the user requested `/foo/bar/` and we can't serve that, we could serve them the contents of `/public/index.html` in the project without rewriting the url. Then, for e.g. a React app, the browser would be served up the contents of `/index.html` (which will correctly load up the React app) and then React would inspect the browser URL - which would be `/foo/bar` - and say, "hey.. that's a route I have" and serve that up. If that was a route the React app did not know about, it would serve up its own 404 behavior.

Hope that's clear.

**Comment by mrubin** [comment]
To add a further note.. in this kind of model, the only reason we need to serve up anything other than `index.html` is the static assets that the React app needs. There is nothing going on in the server otherwise.. no logic other than "serve up a static file we have or serve up the contents of index.html"

(There is a related topic of "server-side rendering" for React apps... but we can leave that out of the discussion for now until we get the basics working.)

**Comment by mrubin** [comment]
Another consideration that I'm now running into while trying to host a React front-end with Play is cache headers. The React build will put some hash characters into the asset filenames, depending on a hash of their contents, as a "browser cache busting" technique. e.g. asset.js will become asset.abc123.js at build time. If the contents of asset.js change in the future, the next time you do a production build it may become asset.456edf.js (so that browsers which may have had asset.abc123.js cached will definitely grab the updated filename from the server).

It may be desirable to set use what Play's `Assets` calls "aggressive caching" on most of the files (the ones which employ this "cache busting" hash). However, index.html cannot change its name so we need to set no-cache headers when serving that file, or serving the contents of it without rewriting the url to satisfy a front-end route.

**Comment by dusanstanojeviccs** [comment]
I think that this is the best approach:

GET /*file  Assets.at(path = "/public", file, fallback = "index.html", fallbackStatus = 200)

It is easy to write for beginners and will offer enough control to serve the SPA index page.

If that is what needs to be done I can probably do it tomorrow.

**Comment by fralken** [comment]
An easy workaround can be implementing a Controller that serves SPA resources like so:

```
@Singleton
class SpaController @Inject()(assets: Assets,
                              cc: ControllerComponents)
                             (implicit ec: ExecutionContext)
  extends AbstractController(cc) {

  def index = assets.at("index.html")

  def assetOrDefault(resource: String) = if (resource.contains(".")) assets.at(resource) else index
}
```
that is, try and find a resource if its name satisfies a condition, otherwise fall back to the index page. Tha condition can be as simple as checking if there is a "." or something more complex such as a regex.

In this way, you get a 404 if trying and getting a non existing resource (such as a missing .css) instead of receiving  "index.html" (as it happens with Nginx with `try_files $uri /index.html`) .

In `routes` file you can add as first route

```
GET        /             controllers.SpaController.index()
```
and, as last route
```
GET        /*file        controllers.SpaController.assetOrDefault(file)
```

**Comment by dusanstanojeviccs** [comment]
@fralken I've been using a similar pattern but without checking for ".":

```
GET     /assets/*file                   controllers.Assets.versioned(path="/public", file: Asset)
GET     /                                      controllers.SPABridge.show()
GET     /*path                             controllers.SPABridge.showWithPath(path: String)
```

**Comment by fralken** [comment]
@dusanstanojeviccs yes but the problem is that "GET /*path" must check whether we are asking for a route or a resource. In case it is a route, it must return "index.html", in case of a resource it must return either the resource or 404 if it doesn't exist.
We must have a criteria to distinguish the two cases. That's mainly the topic of this issue.

**Comment by fralken** [comment]
@dusanstanojeviccs yes I understood your code. I was just considering a slightly different use case where you don't know in advance the path of static resources. In the case you know it (e.g. under "/assets/") of course your implementation is the right way to do it.

**Comment by mrubin** [comment]
> the problem is that "GET /*path" must check whether we are asking for a route or a resource

@fralken I get where you're coming from, but I think this extra logic is unnecessary and also diverges from how nginx behaves. The nginx logic of `try_files $uri /index.html` says: "If I can serve what the user requested, I will. Else, serve up the contents of /index.html (without rewriting the url in the browser)."

If we were asking for a route that exists, the `conf/routes` file should have already matched that route "above". If we make it down to this SPA "catchall" route that's trying to handle the SPA logic, I think we need to assume that we did not match a more-specific route and that we are looking for a resource. Either the resource will be found and served up, or the fallback should be served up.

**Comment by fralken** [comment]
@mrubin I agree with you.

> Either the resource will be found and served up, or the fallback should be served up.

What's the simplest way to implement this? We can use `Assets.at` that is an Action, we should intercept the status 404 result when the resource is not found and return the fallback (index.html) with status 200, but how?

**Comment by mrubin** [comment]
@fralken [This](https://gist.github.com/mrubin/0b6bd0e26bd866681faaa683165f2ec4) is what I've been using. I imagine that it would be possible to have this sort of logic in `Assets` itself, so that you could call something like `Assets.at(..., fallback = "index.html")` or something along those lines.

**Comment by dusanstanojeviccs** [comment]
@fralken @mrubin Hey, so I've got some time tonight and have implemented this logic into Assets themselves, I would really appreciate it if you could take a look at my PR and at the documentation changes and let me know if this would address your use cases. https://github.com/playframework/playframework/pull/8267
